### DISCUSSION TITLE
Action Functions

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#9020] [Official RFCs]
**tl;dr: allow actions to `forms` and `submit` to be client side functions that trigger the usual pending states**

Today actions are configured routes called indirectly through URLs (or the closes route action with an undefined action)

```tsx
let f = <Form action="/some/route/path" method="post" />
submit(data, { action: '/some/route/path', method: 'post' })
```

In the case the action is undefined, it uses the closest route in context.

## Problems

This is mostly fine and matches up with traditional web development, both old-school PHP and SPAs. There was always a route backing a form submission, whether it was another PHP file you posted to or in a SPA, a REST endpoint you called.

Remix took both of those ideas and smooshed it together into one idea, removing a bunch of client/server boilerplate.

**Indirection**

However, there's a level of indirection that can make code harder to track. Seeing a `<Form method="post">` just in the wild, you have to figure out where it's rendered in order to know which action is being called.

**Silliness**

For React Router and Remix SPA actions, it's kinda silly that you have to go through a URL to call a function that's right there, sometimes in the same file. It made a lot of sense for Remix in the beginning where a server was always assumed (and therefore indirection was inherent), but now that we have client-side actions, it's kinda weird.

**Future RPC**

When we first saw SolidStart's `createAction$` we discussed adding it to Remix, but hesitated because it's enabled by some "bundler tricks" that we weren't comfortable with (we weren't doing any AST parsing yet, and as a principle didn't really want to start).

But the DX of the approach is clearly great.

Fast forward, and React has added very Remix like features with "server actions" that are very RPC-like as well with `"use server"` (and SolidStart even use `"use server"` now too). The future is clearly RPC-like actions to add a better DX to calling actions on the server.

In other words, `"use server"` automatically creates the action URL for you, connects the action directly to the component, removing indirection in the code.

So, adding all of this together, the awkwardness of indirection through a URL for client actions, and the future RPC actions, I'd like to propose we get started on this path with a small change.

## Proposal

Anywhere an action takes a URL today, it can also accept a client function (no RPC yet).

```tsx
let f = (
  <Form
    action={async (formData) => {
      // do whatever you want
    }}
  >
    <input type="text" name="title" />
  </Form>
)

submit(
  { title: 'Action Functions' },
  {
    action: async (formData) => {
      // do whatever you want
    },
  },
)
```

Remix router will go through its normal pending states as it waits for the action to finish.

```tsx
let navigation = useNavigation();
navigation.formData?.get("title");)
```

This is probably most useful on fetchers

```tsx
function Comp() {
  let fetcher = useFetcher()
  return (
    <fetcher.Form
      action={async () => {
        // do whatever you want
      }}
    >
      <input name="title" />
      <button disabled={fetcher.state !== 'idle'} />
    </fetcher.Form>
  )
}
```

## Benefits

**Migration to Remix**:

This will make migrating React Router SPAs to Remix SPAs much easier. Most existing React data code is scoped inside of components and hoisting it out to `clientAction` is not straightforward and not always a clear win.

One major issue is how many integrations rely on React Context, for example, `useElements()` from stripe is very difficult to get to work with a clientAction since the stripe hooks and components rely on the same context but `clientAction` isn't in the render tree.

**Removed indirection**:

It's easier to see what code is going to be called when a form is submitted

**Future RPC**:

This skips a step for React Router SPAs path to RSC and server actions. Instead of going from:

- inline data actions
- clientAction
- action
- "use server" action

They can go straight from inline data actions to "use server" actions when Remix supports RPC actions.

- inline data actions
- form action functions
- "use server" actions

## Benefits for Remix SSR

This still benefits Remix SSR apps too and puts people in control of calling their server actions manually if they'd like, while Remix still goes into all of it's pending states as they do:

- Set up a resource route action
- Call it manually from a client action function

```tsx
<Form
  action={async (formData) => {
    // do some client side stuff if you want first
    let response = await fetch('/some/resource/route', {
      method: 'POST',
      body: formData,
    })
    return response.json()
  }}
/>
```

## Future

In the future, Remix can implement `"use server"` and allow people to call server actions directly from their components, skipping the URL indirection and making the code easier to follow.

```tsx
'use server'
export async function someAction() {
  // do some server side stuff
}
```

```tsx
import { someAction } from './someAction'

export function SomeComp() {
  let fetcher = useFetcher()

  return (
    <fetcher.Form action={someAction}>
      <button
        type="submit"
        disabled={fetcher.state !== 'idle'}
      />
    </fetcher.Form>
  )
}
```

Now whether it's a client action or a server action, you wire it up to a form directly.

**Comment by brophdawg11** [comment]
Good news - we have most of the work for this done/reverted in https://github.com/remix-run/react-router/pull/10413

**Reply by brophdawg11** [reply]
see: https://github.com/remix-run/react-router/pull/10362

**Comment by sergiodxa** [comment]
What should happen if I use this to run some client side code and then I call useSubmit?

Since I have no way to await that submit() or fetcher.submit() what would be the pending state?

**Reply by brophdawg11** [reply]
Off the top of my head, an inline `action` is essentially going to behave as a `fetcher.submit` since there's no navigation.  So the inline `action` would complete (and that fetcher would return to `idle`) and the new `submit` call would be reflected in `useNavigation`(or on the second `fetcher` if it was a `fetcher.submit`)

**Reply by ryanflorence** [reply]
can write out a little pseudo code for what you mean @sergiodxa? Want to make sure I catch the nuance.

**Reply by sergiodxa** [reply]
Sure

```ts
let submit = useSubmit()
let location = useLocation();

return (
  <>
    <Form action={action}> inputs here </Form>
    {location.status === "submitting" && <Spinner />}
  </>
)

async function action(formData: FormData) {
  await validateCaptcha(formData);
  submit(formData, { method: "post" })
}
```

Imagine something like this (typically you would use onSubmit to implement it), you need to run some async code like validate a captcha before actually submitting to the server.

With this RFC the `location.status` will be submitting immediately when `action` starts to run, but then since I can't await `submit(formData, { method: "post" })` will location.state go to idle? Will it remain as `submitting` since another submission started? Or will it go to `idle` and then immediately to `submitting` causing a flash on the spinner?

**Reply by sergiodxa** [reply]
So imagine your example where you do a fetch to a resource route, but I want to use `useSubmit` or `fetcher.submit` because instead of a resource route I want to POST to the same route.

**Reply by ryanflorence** [reply]
yeah, probably going to need to return promises from submit, fetcher.submit, fetcher.load for this, dang this has some weirdness we need to think through

```tsx
function Comp() {
  const navigate = useNavigate()
  function action() {
    // do stuff
    navigate() // ?
    // do more stuff but what if I'm unmounted?
  }
}
```

Gotta think through all that, like maybe some things you can't do in an action and it's a no-op with a warning "You can't navigate during an action, throw a redirect"

**Comment by incepter** [comment]
This is good news 🎉🎉🙌
How will the returned value from the action be used?

**Reply by ryanflorence** [reply]
same, `useActionData` and `fetcher.data` none of that changes.

**Comment by sergiodxa** [comment]
Another question, will the `method` change to `POST` when an action function is passed? Or it will still be GET?

And related to that, what will happen before JS loads? Since now the form will need client-side JS to work, maybe this should be added only on `submit` or `fetcher.submit` which are client-side only APIs?

**Reply by brophdawg11** [reply]
> Another question, will the method change to POST when an action function is passed? Or it will still be GET?

I really dislike that API choice in React/Next so I hope we stick with GET 🤞 .  There's no reason the function couldn't behave like a `loader` where the data is serialized into `URLSearchParams`

> maybe this should be added only on submit or fetcher.submit which are client-side only APIs

Agreed - this was the approach we were originally planning on when we first went down this road last year: https://github.com/remix-run/react-router/pull/10362/files

**Comment by NicolasCharpentier** [comment]
_Disclaimer: I don't even use Remix in production (yet), but I've been learning it a lot and loving even much the fact that server entrypoints are all registered at route level._

Would we really get only the `formData`, or would we get the usual `{request, ...}` ? If it is only the `formData` I imagine it would be because of Remix SPA support, but then at the expense of Remix SSR experience and probably promotting bad practices in SSR (like, no security check). Looks dangerous to me then.

**Reply by sergiodxa** [reply]
Remember this only runs client-side, it would be similar to doing:
```ts
<form onSubmit={event => yourAction(new FormData(event.currentTarget))}>
```
So you get the FormData instance, and there's no request yet.

**Reply by NicolasCharpentier** [reply]
oh my bad. Sorry and thank you

**Reply by shamsup** [reply]
This is a good question for future compatibility with server actions though. We would probably want the request in most cases to do things like auth, right?

**Reply by brophdawg11** [reply]
I've been viewing inline action's as an implicit submission to the current location (basically you're providing a route `action` inline instead of on the route), so we could potentially keep the same API and send `request`/`params` if we wanted to.  I'm not as up to date on how the eventual RSC/serverAction APIs will look though so we may choose to align closer to that to ease eventual migrations...

**Reply by tom-sherman** [reply]
Actions accepting a `FormData` as a parameter is something that's specified by React in the native `<form>` props API: https://react.dev/reference/react-dom/components/form#handle-form-submission-on-the-client

This applies to both server and client actions. Of course with a Remix capital F `<Form>` component you can change that API, but that feels confusing as now when you render `<Form>` in a server component it has a different signature compared to on the client. It's also not type safe in TypeScript today.

**Reply by sergiodxa** [reply]
How does React team think devs should check auth if action functions only receive FormData? With a magic function like Next?

**Reply by tom-sherman** [reply]
I can't speak on behalf of them but if you want to read the request then that's a separate concern from React client & server actions. React is agnostic of the transport protocol of your actions so it can't pass you the request object. It's up to the framework to do that.

**Reply by sergiodxa** [reply]
> React is agnostic of the transport protocol of your actions

Ok this actually makes a lot of sense, I suppose frameworks should either do the Next approach of a magic function or just augment the arguments of the server action to include extra data.

Maybe with middlewares and server context it should be possible to inject the server context like `declare action(formData: FormData, context: ServerContext): Promise<void> | void` and use server context to pass things from the request like user session.

**Reply by tom-sherman** [reply]
I don't think the extra params would work with the native `<form>` element but I'm not 100% sure about that.

**Reply by sergiodxa** [reply]
The React docs suggests using `action.bind` to add extra arguments, even for server actions, so Remix could just do that, maybe move context to before formData.

**Reply by tom-sherman** [reply]
`.bind()` round trips the extra args via hidden form elements so probably not suitable for the auth use case.

**Comment by romansndlr** [comment]
Will any loaders be called after the action has run?

**Reply by brophdawg11** [reply]
yeah the behavior of actions and revalidation won't change

**Comment by ZeldOcarina** [comment]
While this seems a great dev experience, my only concern would be how it abides to the Remix philosophy of using web standards as much as possible…

**Comment by michalwarda** [comment]
I think it's a good approach for Client based actions.

For server actions (future) I think the url approach is much better when it comes to a few aspects, but the main one is obviously scope. What Next is doing and React is going for lately is incoherent with language behavior. I personally have a lot of trouble explaining the boundary between server and client to junior devs in the new approach used ie. in Next. Remix is much simpler. This function is a server one, this is a client one. They are separated with files now even (thanks to new Vite bundler and `.server`). 

Onboarding new devs to our Remix variant of app is so much easier. It takes around 1/4th of the time it took us in Next. Also good patterns like dependency injection and separation of concerns are very simple to explain and implement. Testing is also straightforward and we're not forced to do e2e testing everywhere. All of that thanks to the fact that there is a clear (url) boundary between those two contexts.

I remember a few years back the big reason for moving away from class components and why it was happening was that users could not understand what _binding variables was_. `this` was feared. We got order based hooks as an alternative to that. No more _this_. But because of that we lost the possibility to use if statements where ever we wanted in components because hooks order broke the "react vm". I think server functions will do a very similar thing. We will lose more of the basic JS features like scope and React will continue to move away from JavaScript (or TS).

So besides the rant above :D, I'm wondering if this proposal will do any context reconstructing for client functions?

ie.
 
 ```
 const a = 5;
 ...
 <Form
  action={async (formData) => {
    // do some client side stuff if you want first
    let response = await fetch(`/some/resource/route/${a}`, {
      method: 'POST',
      body: formData,
    })
    return response.json()
  }}
/>
```

The `a` variable will be just available? Maybe there should be also be some "runtime" check disabling referencing those for the future server actions so that users would not have to rewrite them later. Maybe a second parameter?

 ```
 const a = 5;
 ...
 <Form
  action={async (formData, context) => {
    // do some client side stuff if you want first
    let response = await fetch(`/some/resource/route/${context.a}`, {
      method: 'POST',
      body: formData,
    })
    return response.json()
  }}
  context={{a}}
/>
```

**Reply by shamsup** [reply]
The Future example shows importing the server action for remix SSR/RSC, so it looks like we would keep the server and client code separated still.

**Comment by andrelandgraf** [comment]
What happens if the route module also exports a `clientLoader` or `clientAction`?

**Reply by brophdawg11** [reply]
The route client-side action is either:
* A `fetch` to the server `action` if only a server `action` is provided
* The `clientAction` function which gives you `serverAction` to call the server if you want to

I think an inline action function would become the client-side action and override both of the above scenarios.  I don't think we'll want to expose it the concept of `serverAction` since this is intended as a bridge to eventual `"use server"`.  As mentioned above, if you wanted to talk to the server, you would do it via a resource route:

```jsx
<Form
  action={async (formData) => {
    // do some client side stuff if you want first
    let response = await fetch('/some/resource/route', {
      method: 'POST',
      body: formData,
    })
    return response.json()
  }}
/>
```

So you could either use `action`/`clientAction`(today) or a resource route/inline action (after this RFC)

**Comment by SteveSuv** [comment]
Can be an API similar to `useMutation` in `react-query`?
Then the action doesn't have to bind with `form`.

```tsx
export const loader = () => {
  return axios.get("/islogin");
};

export default () => {
  const { data: islogin } = useLoaderData<typeof loader>();

  const { isLoading, mutateAsync } = useAction(({ username, password }) => {
    return axios.post("/login", { username, password });
  });

  if (islogin) {
    return <div>You are already logged in</div>;
  }

  // use Form
  return (
    <Spin loading={isLoading}>
      <Form
        action={async (formData) => {
          try {
            // mutateAsync will auto revalidate loader data if the action promise resolved
            const res = await mutateAsync(formData);
            if (res.data) {
              toast.success("login successful");
            }
          } catch (err) {
            console.error(err);
            toast.error("wrong username or password");
          }
        }}
      >
        {/* ... */}
      </Form>
    </Spin>
  );

  // or not use Form
  return (
    <Button
      loading={isLoading}
      onClick={async () => {
        try {
          // mutateAsync will auto revalidate loader data if the action promise resolved
          // username and password not from formData but something like useState 
          const res = await mutateAsync({ username, password });
          if (res.data) {
            toast.success("login successful");
          }
        } catch (err) {
          console.error(err);
          toast.error("wrong username or password");
        }
      }}
    >
      login
    </Button>
  );
};
```

**Comment by ryanflorence** [comment]
I now expect https://github.com/remix-run/react-router/issues/11564 to remove the need for this, closing for now. If I'm wrong we can revisit this.
